subsys/bluetooth/mesh/friend.c

```diff
@@ -962,6 +937,10 @@ static void enqueue_friend_pdu(struct bt_mesh_friend *frnd,
 		}
 
 		sys_slist_merge_slist(&frnd->queue, &seg->queue);
+		seg->seg_count = 0U;
+	} else {
+		/* Mark the buffer as having more to come after it */
+		buf->flags |= NET_BUF_FRAGS;
```

Comment:
Wouldn't this create a NULL pointer reference net_buf_slist_get if we purge the queue before more segments can be added? The for loop in net/buf.c:472 seems to trust this flag enough to not do null checks

Reply:
@trond-snekvik yes, that's a good catch. Didn't think about this when I started taking advantage of the net_buf fragmentation feature (which really is almost the same thing as the segments, except that the segments come gradually and we might end up with an incomplete set as you point out).

Reply:
I'm now wondering should we move to something else than NET_BUF_FRAGS or should we fix buf.c to handle incomplete data in the FIFO.

Reply:
This really only affects the segment list, and not the friend queue, since we only move complete messages to the friend queue. I'm thinking since we mess around with what's essentially an net_buf-internal flag, we should just do the purging ourselves rather than using net_buf helpers.

Reply:
@trond-snekvik so I'm thinking to just do this:
```
--- a/subsys/bluetooth/mesh/friend.c
+++ b/subsys/bluetooth/mesh/friend.c
@@ -110,7 +110,14 @@ struct bt_mesh_friend *bt_mesh_friend_find(u16_t net_idx, u16_t lpn_addr,
 static void purge_buffers(sys_slist_t *list)
 {
        while (!sys_slist_is_empty(list)) {
-               net_buf_unref(net_buf_slist_get(list));
+               struct net_buf *buf;
+
+               buf = (void *)sys_slist_get_not_empty(&frnd->queue);
+
+               buf->frags = NULL;
+               buf->flags |= ~NET_BUF_FRAGS;
+
+               net_buf_unref(buf);
        }
 }
```